### ISSUE TITLE
Doubts about repacked tables relfrozenxid

### ISSUE THREAD

**Issue by dvarrazzo** [reorg/pg_repack#152]
I'm down in the rabbit hole trying to understand #135. Looking at what the `swap_head_or_index_files()` does [here](https://github.com/reorg/pg_repack/blob/ver_1.4.1/lib/repack.c#L1131), it seems that the original author wanted to set the repacked table's relfrozenxid to the greatest of the two, but the code does the opposite (or better it sets the relfrozenxid on the temporary table that then it drops).

```
contrib_regression=# select  relfrozenxid, txid_current() from pg_class where oid = 'test'::regclass;
 relfrozenxid | txid_current 
--------------+--------------
       444833 |       444901
(1 row)

contrib_regression=# vacuum full test;
VACUUM
contrib_regression=# select  relfrozenxid, txid_current() from pg_class where oid = 'test'::regclass;
 relfrozenxid | txid_current 
--------------+--------------
       444902 |       444903
(1 row)

contrib_regression=# cluster test using test_pkey;
CLUSTER
contrib_regression=# select  relfrozenxid, txid_current() from pg_class where oid = 'test'::regclass;
 relfrozenxid | txid_current 
--------------+--------------
       444905 |       444906
(1 row)
```

What `pg_repack` does is sort of backwards:

```
contrib_regression=# \! /usr/lib/postgresql/9.6/bin/pg_repack -d contrib_regression -t test
INFO: repacking table "test"

# adding a couple of LOG in the code:
LOG:  relform1->relfrozenxid = 444905
LOG:  relform2->relfrozenxid = 444908

contrib_regression=# select  relfrozenxid, txid_current() from pg_class where oid = 'test'::regclass;
 relfrozenxid | txid_current 
--------------+--------------
       444905 |       444917
(1 row)
```

shouldn't we set the repacked table frozen xid to the one of the new table?

**Comment by schmiddy** [comment]
You may already know all this @dvarrazzo , but just to summarize some code archaelogy from a little digging I did: the comment above our `swap_heap_or_index_files ` [says](https://github.com/reorg/pg_repack/blob/4b86bb1c4117e56d2f69418dccfb7b46892c29fd/lib/repack.c#L1083-L1084):

> This is a copy of swap_relation_files in cluster.c, but it also swaps relfrozenxid.

Looks like way back in the days of PG 8.2 , the signature was: [`swap_relation_files(Oid r1, Oid r2)`](https://github.com/postgres/postgres/blob/105f3ef492abd22985321b64e8f40e58f3a15a83/src/backend/commands/cluster.c#L793). And then for PG 8.3, that changed to [`swap_relation_files(Oid r1, Oid r2, TransactionId fronzenXid`](https://github.com/postgres/postgres/blob/d13f41d2153890bf817451c152a83d2867e61199/src/backend/commands/cluster.c#L928).  I bet the original author copied from PG 8.2, and added in this logic for relfrozenxid copying which maybe he mixed up.  I think we should try to update our `swap_relation_files()` signature to mimic the slightly-more-recent version of cluster.c from PG. (And at some point we should catch up to the even more recent changes to `swap_relation_files()`...)

**Comment by dvarrazzo** [comment]
Yes I noticed the fork happened between 8.2 and 8.3. We can go forward version-by-version I guess.

The function signature in the latest version has sprouted a lot of arguments...

**Comment by bashtanov** [comment]
I'm sorry I noticed this conversation only after I've opened https://github.com/reorg/pg_repack/pull/157.

Anyway, why is it a bad idea to swap `relfrozenxid`s together with the data? It looks correct, as table's `relfrozenxid` is consistent with its data.

For the new table, it's actually the transaction the table was created, unless the table was autovacuumed in the meanwhile, which I doubt repack transactions and locks would allow to happen. Even if it would, autovacuum would update the `relfrozenxid` correctly. So it looks like this value is good enough and we cannot use any later one instead.

**Comment by bashtanov** [comment]
@dvarrazzo @schmiddy I would appreciate your opinion on my comment and PR

**Comment by dvarrazzo** [comment]
Hi @bashtanov sorry for not getting back to you.

My intention is to work on this problem and on #135 in a few days, during the xmas break. I'll look into your MR but I guess it will be "swallowed" by a sea of other changes aligning the `swap_head_or_index_files()`  to the current state in Postgres code.

**Comment by MichaelDBA** [comment]
Where are we on this issue and the underlying issue that is also still open related to a problem with logical replication and pg_repack (#135)?  I haven't seen an update in a few months.  Also I see a subsequent issue that is related to this one as well (#157).

I hesitate to use pg_repack anymore in my v10 logical replication environment.

**Comment by dvarrazzo** [comment]
@MichaelDBA We are discussing just now about it in #157, as you have probably read (so why your message here, even?).

The issue (which really is [this](https://www.postgresql.org/message-id/CA%2Bmi_8YWReON2gVk9qoeJLRVzWTSSoLkKAX9DERL36-n4Y8rZg%40mail.gmail.com), #152 is only marginal) is complex and is about postgres internals evolution: the people working in this project have only a marginal understanding of what happened in the 10 years after this code was written and my pledge for help has gone basically ignored. I can speak only for myself: I have limited time for unpaid work, mine is volunteering. If work will be done, you will benefit of it for free, as you have done so far. If you want to contribute to the problem in a more constructing way than "I hesitate to use pg_repack anymore" you are welcome. If you are a core Postgres developer then please help fixing the problem or convince (pay, lure, threaten) someone to do it.

**Comment by MichaelDBA** [comment]
Thanks for the comments, Daniele.  I guess what I was interested in is what thread to follow on this issue.  Couldn't we just close #152 and #135 saying it is being resolved through issue, #157?

That way we don't have to monitor multiple issues kinda dealing with the same problem.

**Comment by dvarrazzo** [comment]
@MichaelDBA 

#157 is a pull request which fixes #152, which is a bug about xid swapped. Pretty much standard github workflow.

#135 is a bug that says logical decoding is broken, which might be fixed updating the code to the current architecture and is not closed by #157.

Talking with the Postgres core people is a disheartening experience I have gone through several time already: I am not at their level, both technically and masochistically. I keep on doing it if I feel the project at hand needs it, but TBH I have better used of my time than being publicly called a moron, so I save the taste for the special occasions. Yes, I believe this project shouldn't even exist and postgres should defrag its table by itself, but they all seem fine the way with the way things are (_Bloat? What bloat?_) and I don't think I am able to talk them into it.
